test/utils/test_common.py

```diff
@@ -70,3 +76,90 @@ def test_int(self, expected):
     def test_tuple(self, expected):
         actual = to_tuple_tuple((2, 2, 2), 3, 3)
         assert actual == expected, "tuple -> tuple[tuple] incorrect"
+
+
+class TestCheckpointWrapper:
+    @pytest.fixture
+    def device(self):
+        return torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+
+    @pytest.fixture
+    def model(self, device):
+        class DummyAttention(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.param = nn.Parameter(torch.ones(3))
+
+            def forward(self, x, y, attn_mask=None, use_cache=None):
+                u = x if use_cache else y
+                u = u * self.param
+                if attn_mask is not None:
+                    u = u * attn_mask
+                return u
+
+        class DummyModel(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.attention = DummyAttention()
+                self.identity = nn.Identity()
+
+            @checkpoint_wrapper
+            def _layer_to_wrap(self, x, y, attn_mask, use_cache):
+                return self.attention(x, y, attn_mask, use_cache)
+
+            def forward(self, x, y, attn_mask=None, use_cache=False):
+                out = self._layer_to_wrap(x, y, attn_mask, use_cache=use_cache)
+                out = self.identity(out)
+                return out
+
+        return DummyModel().to(device)
+
+    @pytest.fixture
+    def inputs(self, device):
+        # Set inputs to be requires_grad = True or else it will break autograd
+        # during the backward pass,
+        # "RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn"
+        x = torch.ones(3, requires_grad=True)
+        y = torch.ones(3, requires_grad=True) * 2
+        attn_mask = torch.tensor([1, 1, 0])
+        x = x.to(device)
+        y = y.to(device)
+        attn_mask = attn_mask.to(device)
+
+        return x, y, attn_mask
+
+    @pytest.fixture
+    def compute_grad(self, model):
+        model.zero_grad()
+
+        def _compute_grad(output):
+            output.sum().backward()
+            grad_checkpointed = {}
+            for name, param in model.named_parameters():
+                grad_checkpointed[name] = param.grad.data.clone()
+            return grad_checkpointed
+
+        return _compute_grad
+
+    def test_training_mode(self, model, inputs, compute_grad):
```

Comment:
So in these two tests it seems you are checking that use_cache gets disabled in train mode, but are you actually testing that the checkpointing gets applied in some way? Btw might be nice to add a quick comment explaining why the grad is the same as the model output, I was a bit confused by this at first glance.

Reply:
Set `use_cache=False` is in the same codepath as `torch.utils.checkpoint.checkpoint` in a way that the former is a proxy of the latter. But I can see how this may be less obvious in regard to what's being tested. I have not seen a good way of testing triggering of `checkpoint` API elsewhere. 

I can also patch `checkpoint` and inspect the mock that it's called.

Reply:
Thanks for clarifying, I think that makes sense. Up to you if you want to patch checkpoint or not, but if not I would at least add a comment explaining this point.